src/cmdstan/command.hpp

```diff
@@ -517,300 +327,195 @@ int command(int argc, const char *argv[]) {
-        output_stream << gradients.back() << "\n";
-      }
-      return stan::services::error_codes::error_codes::OK;
+      services_log_prob_grad(model, jacobian, params_r_ind, sig_figs,
+                             sample_writers[0].get_stream());
+      return_code = return_codes::OK;
     } catch (const std::exception &e) {
-      return stan::services::error_codes::error_codes::DATAERR;
+      return_code = return_codes::NOT_OK;
     }
+    //////////////////////////////////////////////////
   } else if (user_method->arg("diagnose")) {
     list_argument *test = dynamic_cast<list_argument *>(
         parser.arg("method")->arg("diagnose")->arg("test"));
-
     if (test->value() == "gradient") {
-      double epsilon
-          = dynamic_cast<real_argument *>(test->arg("gradient")->arg("epsilon"))
-                ->value();
-      double error
-          = dynamic_cast<real_argument *>(test->arg("gradient")->arg("error"))
-                ->value();
+      double epsilon = get_arg_val<real_argument>(*test, "gradient", "epsilon");
+      double error = get_arg_val<real_argument>(*test, "gradient", "error");
       return_code = stan::services::diagnose::diagnose(
           model, *(init_contexts[0]), random_seed, id, init_radius, epsilon,
           error, interrupt, logger, init_writers[0], sample_writers[0]);
     }
+    //////////////////////////////////////////////////
   } else if (user_method->arg("optimize")) {
+    int num_iterations
+        = get_arg_val<int_argument>(parser, "method", "optimize", "iter");
+    bool save_iterations = get_arg_val<bool_argument>(
+        parser, "method", "optimize", "save_iterations");
+    bool jacobian
+        = get_arg_val<bool_argument>(parser, "method", "optimize", "jacobian");
     list_argument *algo = dynamic_cast<list_argument *>(
         parser.arg("method")->arg("optimize")->arg("algorithm"));
-    int num_iterations = dynamic_cast<int_argument *>(
-                             parser.arg("method")->arg("optimize")->arg("iter"))
-                             ->value();
-    bool save_iterations
-        = dynamic_cast<bool_argument *>(
-              parser.arg("method")->arg("optimize")->arg("save_iterations"))
-              ->value();
-
     if (algo->value() == "newton") {
-      return_code = stan::services::optimize::newton(
-          model, *(init_contexts[0]), random_seed, id, init_radius,
-          num_iterations, save_iterations, interrupt, logger, init_writers[0],
-          sample_writers[0]);
+      if (jacobian)
+        return_code
+            = stan::services::optimize::newton<stan::model::model_base, true>(
+                model, *(init_contexts[0]), random_seed, id, init_radius,
+                num_iterations, save_iterations, interrupt, logger,
+                init_writers[0], sample_writers[0]);
+      else
+        return_code
+            = stan::services::optimize::newton<stan::model::model_base, false>(
```

Comment:
good idea.  can that be generalized for log_prob and laplace as well?

Reply:
I think so, yes. It wouldn’t make the repeated pattern any more concise, but at least shuffles it to one central place

Reply:
to generalized, how would I pass in the service method to the inline helper function?  as an argument or template parameter?

Reply:
I don’t know enough about the variadic templating/arguments in C++. I *think* you can write something like 

```cxx
template <typename F, typename Ts…> auto jacobian_wrapper(F f, bool jacobian, Ts args…)
```

but I have no idea if this will work off the bat or if it is just easier to write a new function for each.

If it’s not possible as a template, it probably is as a macro, but I’d avoid that I think

Reply:
Yes, you can use [parameter packs](https://en.cppreference.com/w/cpp/language/parameter_pack) to write variadic arguments recursively.  Here's a [tutorial with simple examples](https://kevinushey.github.io/blog/2016/01/27/introduction-to-c++-variadic-templates/).

Reply:
understand parameter packs - question is w/r/t function argument - I can just pass around function methods as args and invoke them?

```
if (jacobian) 
 return f<model, true>(Args&& args);
else
 return f<model, false>(Args&& args);
```

Reply:
The expression`f<model, true>` doesn't make sense with `f` being an instance.   Every instance has to be concrete and specify all parameters.  So you have to make this decision higher up the call chain before `f` is instantiated or you have to be lazy and not instantiate `f` until you get inside the `if (jacobian)` conditional.   

I would suggest turning `jacobian` into a template parameter at a scope that includes `f`'s instantiation.  You can see that's what I did for the service functions for optimization.

Reply:
I just tried for a while on godbolt and couldn't find a version which let me write down a template for `f`. 
https://godbolt.org/z/e6bje4ja6

It may still be possible, just my first attempt at writing it didn't work

Reply:
You can't put template parameter on instances.  You can create a struct with a template and pass that in as follows.

```cpp
template <bool disable>
struct foo {
  int operator()(int x, double y) {
    return disable ? 0.0 : x * y;
  }
};

template <template <typename> class F, bool disable>
auto wrap_disable(Ts... xs) {
    F<disable> f;
    return f(xs...);
}
```

It pushes the construction of `F` done inside the `wrap_disable` function.  This will only work if the argument `F` can work with the single argument.

Maybe if you said what you're trying to do, we can work backward from there.

Reply:
We have a bunch of places now where we have code that looks like

```cpp
if (jacobian)
    call_some_function<true>(a, bunch, of, args);
else
    call_some_function<false>(the,same,args,here);
```

My original inquiry was if there's a way to write a helper which handles this logic for us. However, I am beginning to suspect that the cure would be far worse than the original disease looking at the examples of code which does similar things

Reply:
this should work if that single template parameter is a parameter pack, no?

Reply:
>  This will only work if the argument F can work with the single template parameter.

Yeah, I don't thing we'd want to turn the functions we're talking about here into structs (they're the services functions, mainly), _and_ some of them are called with `<model, false>` (e.g. more than one template parameter), while others are only called with the boolean for the jacobian adjustment.

I think I've seen enough to declare it's not worth trying to write a one-size-fits wrapper. That being said, we may want to consider writing (manually) some overloads for the service functions which wrap up this dispatch

Reply:
now I understand the problem.   no can do on this one.